testr/packages.py

```diff
@@ -273,17 +300,16 @@ def run_tests(package, tests):
                 else:
                     test['status'] = 'pass' if sub.returncode == 0 else 'FAIL'
             else:
-                if interpreter == 'bash':
-                    with open(test['file'], 'r') as fh:
-                        cmd = fh.read()
-                elif interpreter is None:
-                    cmd = './' + test['file']
+                if interpreter is None:
+                    cmd = [f"./{test['file']}"]
                 else:
-                    cmd = interpreter + ' ' + test['file']
+                    cmd = [interpreter, test['file']]
 
-                try:
-                    bash(cmd, logfile=logfile, env=env)
-                except ShellError:
+                process = subprocess.Popen(
+                    cmd, env=env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
+                )
+                communicate(process, log=logfile)
+                if process.returncode:
```

Comment:
I'm not sure if this is equivalent to the original `ShellError`. I can think of three ways for something to fail:
1. Command never starts, i.e. you give a `cmd` that does not exist. I think that this will raise an exception in the `subprocess.Popen` which is not currently caught.
2. Command starts but crashes at some point, i.e. an exception in Python code.
3. Command completes but returns a status != 0.  (i.e. `sys.exit(1)` in Python. This might be the same as (2) but I'm not sure.

In any case you should test all three of these possibilities.

Reply:
It would probably make testing easier if you factor out this code into a local function `bash` with a similar API as the original (meaning it raises a well-defined exception if anything goes wrong, otherwise not). Then it will be trivial to check those cases and be confident that this is not changing things.

Reply:
Actually, the easiest is what I just did. I wrote this little program I called `fail.py`:
```python
#!/usr/bin/env python

import sys
if sys.argv[1:]:
    rc = int(sys.argv[1])
    sys.exit(rc)
sys.exit(0)
```

1. `bash('asd')`: 
   ```
   ShellError: Shell command asd failed with exit status 127
   ```
3. `bash('./fail.py asd')`: 
   ```
   ShellError: Shell command ./fail.py asd failed with exit status 1
   ```
5. `bash('./fail.py 2')`:
   ```
   ShellError: Shell command ./fail.py 2 failed with exit status 2
   ```

Summary, they all raise `ShellError` and it can always be summarized by the return code.

Reply:
If you run subprocess and the command does not exist, you get a `FileNotFoundError` exception earlier (in line 38).

Reply:
It looks like you have verified the API of the `Ska.Shell` version of `bash` in the 3 cases, for which the original code in `testr` generates a FAIL status.

What I was hoping for is for explicit verification by test (functional or unit) that the new code in `testr.packages.run_tests()` also returns a FAIL in those 3 cases. You can do that any way you like, but if it were me since the new logic is buried deep in `run_tests`, I would find it easier to factor out the key bits and write unit tests.

(BTW I haven't looked into the existing tests to see if this is already done. If so then ignore this.)

Reply:
I must be missing something. The code is basically:
```python
process = subprocess.Popen(
    cmd, env=env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
)
...
if process.returncode:
    test['status'] = 'FAIL'
```
To me it seemed so obvious that this is correct (i.e.: pytest returns RC>0 on failure, subprocess returns RC>0 on missing file or any other failure, this is all standard) that I answered to the question "this is equivalent to the original ShellError" by checking the "original ShellError". 

Basically, testr itself has no tests. Zero. Are you asking that I add unit tests testr in general?